vortex-buffer/src/buffer.rs

```diff
@@ -0,0 +1,218 @@
+pub unsafe trait ByteSource {
+    /// The type of the owner of the bytes.
+    type Owner: ByteOwner;
+
+    /// Returns the bytes as a slice.
+    fn as_bytes(&self) -> &[u8];
+
+    /// Returns the raw owner of the bytes.
+    fn into_owner(self) -> Self::Owner;
+
+    /// Attempts to convert this buffer into a mutable buffer.
+    fn into_mut(self: Arc<Self>) -> Result<BufferMut, Arc<dyn ByteOwner>>;
+}
+
+/// A trait for types that own the bytes.
+pub trait ByteOwner: Sync + Send + 'static {
+    /// Downcasts the owner to `Any`.
+    fn as_any(self: Arc<Self>) -> Arc<dyn Any + Sync + Send>;
+
+    /// Attempts to convert this buffer into a mutable buffer.
+    fn into_mut(self: Arc<Self>) -> Result<BufferMut, Arc<dyn ByteOwner>>;
+}
+
+impl<T: ByteSource + Sync + Send + 'static> ByteOwner for T {
+    #[inline(always)]
+    fn as_any(self: Arc<Self>) -> Arc<dyn Any + Sync + Send> {
+        self
+    }
+
+    #[inline(always)]
+    fn into_mut(self: Arc<Self>) -> Result<BufferMut, Arc<dyn ByteOwner>>
+    where
+        Self: Sized,
+    {
+        <T as ByteSource>::into_mut(self)
+    }
+}
+
+impl Buffer {
+    /// Creates a new empty `Buffer`.
+    pub fn empty() -> Self {
+        Self::from_owner(&[0u8; 0][..])
+    }
+
+    /// Creates `Buffer` from a [`ByteOwner`] + [`ByteSource`].
+    pub fn from_owner(owner: impl ByteSource + ByteOwner) -> Self {
+        let owner = Arc::new(owner);
+        let slice = owner.as_bytes();
+        let ptr = slice.as_ptr();
+        let length = slice.len();
+        Self { owner, ptr, length }
+    }
+
+    /// Creates `Buffer` from an `Arc<ByteSource + ByteOwner>`.
+    pub fn from_owner_arc(arc: Arc<impl ByteSource + ByteOwner>) -> Self {
+        let slice = arc.as_bytes();
+        let ptr = slice.as_ptr();
+        let length = slice.len();
+        Self {
+            owner: arc,
+            ptr,
+            length,
+        }
+    }
+
+    /// Extract and downcast the owner of the bytes.
+    /// TODO(ngates): does this API make it too easy to forget about sliced data?
+    pub fn into_owner<T>(self) -> Option<Arc<T>>
+    where
+        T: Send + Sync + 'static,
+    {
+        self.owner.as_any().downcast::<T>().ok()
+    }
+
+    /// Attempts to convert this buffer into a mutable buffer if there is only a single
+    /// reference to the buffer. Otherwise, returns the original buffer.
+    pub fn try_into_mut(self) -> Result<BufferMut, Self> {
+        self.owner
+            .into_mut()
+            .map(|owner_mut| BufferMut {
+                owner: owner_mut.owner,
+                ptr: self.ptr as *mut u8,
+                length: self.length,
+            })
+            .map_err(|owner| Self {
+                owner,
+                ptr: self.ptr,
+                length: self.length,
+            })
+    }
+
+    /// Returns a slice of self for the provided range.
+    /// This operation is `O(1)`.
+    pub fn slice(&self, offset: usize) -> Self {
```

Comment:
slice takes an offset instead of a range. this seems wrong?

Reply:
This is the Arrow API, I actually prefer it. The example below is a Flatbuffers trait which is why it's a bit weird.